{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"1s-count->TPM.ipynb","provenance":[],"collapsed_sections":[],"mount_file_id":"1T7DF71yFOA3ePsApzpDM77kSpB7lx_RE","authorship_tag":"ABX9TyMKh+scvK67aUS0VoZQNqRY"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","metadata":{"id":"LWVxOAY8Mg6Z","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1638250382396,"user_tz":-540,"elapsed":246,"user":{"displayName":"久内佑樹","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gg8aqObfU0mdL-poY9eo71yADwG-v3GfTCsl8fn=s64","userId":"01891010281923395733"}},"outputId":"04033ab9-c654-4192-9682-1df9e02cd9af"},"source":["#@title ドライブをマウント \n","from google.colab import drive\n","drive.mount('/content/drive')"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n"]}]},{"cell_type":"code","metadata":{"id":"8tA2tvq-FFZ5"},"source":["#@title データを保存するディレクトリのパスを指定\n","#@markdown ランタイムに接続しドライブを後、コピーしてくる<br>\n","#@markdown 後ろの / は追加するので、追記は不要<br>\n","\n","cwd = \"/content/drive/MyDrive/RNAseq/Outputs/Hypoxia2\" #@param {type:\"string\"}\n","cwd = cwd + \"/\"\n","\n","# 必要なモジュールをインポートする\n","import pandas as pd\n","import numpy as np\n","import urllib.parse\n","from google.colab import files\n","\n","# カレントディレクトリの変更\n","import os\n","os.makedirs(cwd , exist_ok = True)\n","os.chdir(cwd)\n","currentdir = os.getcwd()"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"_5ZekMN5FlNu"},"source":["#@title 実験名とN数を設定\n","#@markdown 保存ファイル名の前に付ける名前を設定する<br>\n","#@markdown N数は1と3のみ実装、他の数の場合は分けて適宜使用<br>\n","exp_name = \"DC_only\" #@param {type:\"string\"}\n","exp_name = exp_name + \"_\"\n","n_number = \"3\" #@param [\"1\", \"3\"]"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"_baAybUcGYQk"},"source":["#@title 各ファイルのパスを指定\n","#@markdown N=1の場合は、カウントデータ2, 3を空白にする\n","\n","#@markdown コントロール群のカウントデータ(N=1, N=3の1つ目)\n","count_Cont1 = \"/content/drive/MyDrive/RNAseq/Inputs/Hypoxia2/F45_NRX_DC.txt\" #@param {type:\"string\"}\n","#@markdown コントロール群のカウントデータ2(N=3の2つ目)\n","count_Cont2 = \"/content/drive/MyDrive/RNAseq/Inputs/Hypoxia2/M21_NRX_DC.txt\" #@param {type:\"string\"}\n","#@markdown コントロール群のカウントデータ3(N=3の3つ目)\n","count_Cont3 = \"/content/drive/MyDrive/RNAseq/Inputs/Hypoxia2/M22_NRX_DC.txt\" #@param {type:\"string\"}\n","\n","#@markdown テスト群のカウントデータ1(N=1, N=3の1つ目)\n","count_Test1 = \"/content/drive/MyDrive/RNAseq/Inputs/Hypoxia2/F45_HYP_DC.txt\" #@param {type:\"string\"}\n","#@markdown テスト群のカウントデータ2(N=3の2つ目)\n","count_Test2 = \"/content/drive/MyDrive/RNAseq/Inputs/Hypoxia2/M21_HYP_DC.txt\" #@param {type:\"string\"}\n","#@markdown テスト群のカウントデータ3(N=3の3つ目)\n","count_Test3 = \"/content/drive/MyDrive/RNAseq/Inputs/Hypoxia2/M22_HYP_DC.txt\" #@param {type:\"string\"}\n","\n","#@markdown アノテーションファイル\n","gtf_path = \"/content/drive/MyDrive/RNAseq/SourceFiles/Homo_sapiens.GRCh38.94.gtf\" #@param {type:\"string\"}"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"zHXFZ4gbHXuW"},"source":["#@title featurecountsの出力名が長いので、サンプル名をわかりやすく変更する\n","#@markdown 以下に、変更後の名前を指定する\n","Cont_name = 'N_DC' #@param {type:\"string\"}\n","Test_name = 'H_DC' #@param {type:\"string\"}"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"5jVpk3kkcuH8"},"source":["ここまでを入力後、「ランタイム」タブ→「すべてのセルを実行」を選択(ショートカット：Ctrl + F9)<br>\n","データはGoogleドライブ上に保存される"]},{"cell_type":"markdown","metadata":{"id":"k3yMCpc7Hhk9"},"source":["===================================================================\n","============================================"]},{"cell_type":"code","metadata":{"id":"ZKdayQmsHg-v"},"source":["#@title gene_idとproductの対応表作成\n","#@markdown アノテーションファイルから、Gene_idとproduct(Gene_name)の対応表を作成する<br>ファイルの形式によっては、コードを多少書き換える必要である場合もある<br>ensemblの出してるgtf, gffファイルならこのコードでいけるはず\n","\n","# 出力用のリストを定義\n","outputlist = []\n","# 出力用のcsvを作成\n","fout = open(exp_name + '_id_name.csv', 'w')\n","# gtfファイルを開く\n","with open(gtf_path, 'r') as fin:\n","    # gtfデータを1行ずつ読み込む\n","    for line in fin:\n","      # 出力用の変数を定義\n","        product = ''\n","        geneid = ''\n","        \n","        # 行末の空白を削除\n","        line=line.rstrip()\n","        # #で始まっているデータ行を除く\n","        if not line.startswith('#'):\n","            # gtfをタブ区切りで認識し、その9列目(1列目が0なので8)を読む\n","            s=line.split('\\t')\n","            items=s[8].split(';')\n","            \n","            # 各行の9列目について\n","            for item in items:\n","                # ' gene_name \"'で始まる要素なら、\"で分割して=の右側を取出し、url逆変換する\n","                if item.startswith(' gene_name \"'):\n","                    product = item.split('\"')[1]\n","                    product = urllib.parse.unquote(product)\n","\n","                # 'gene_id \"'で始まる要素なら、\"で分割して右側を取出す\n","                if item.startswith('gene_id \"'):\n","                    geneid = item.split('\"')[1]\n","\n","                # gene_name と gene_idが両方あれば、出力リストに取り込む\n","                if (geneid != '') and (product != ''):\n","                    outputlist.append(geneid+','+product)\n","\n","# 重複を除去、並び順をソートの上で、ヘッダーを追加する\n","outputlist = sorted(list(set(outputlist)))\n","outputlist.insert(0 , \"Geneid\"+','+\"product\")\n","  \n","# csvに書き込んでいく\n","fout = open(exp_name + \"id_name.csv\", \"w\")\n","for u in outputlist:\n","    print(u, file=fout)\n","fout.close()\n","\n","# csvをData frameとして認識後、ヘッダーを除去して再保存する\n","df_id_name = pd.read_csv(exp_name + '_id_name.csv')\n","df_id_name.to_csv(exp_name + '_id_name.csv' ,sep = ',' , index=False , header=False) \n","\n","# 対応表のData Frameを表示する\n","\n","# print(df_id_name)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"eJiJJnyWbl5Z"},"source":["===================================================================\n","============================================"]},{"cell_type":"code","metadata":{"id":"gamMk7ukKSbr"},"source":["#@title カウントデータの調整\n","#@markdown カウントデータから必要な部分を抽出し、Gene_idを元に結合する<br>必要な部分：Geneid , Length , コントロール群のカウントデータ , サンプル群のカウントデータ\n","\n","# N=1の処理\n","if n_number == \"1\":\n","  # データの読み込み\n","  df_Cont1 = pd.read_table(count_Cont1 , index_col = 0 , skiprows = 1)\n","  df_Test1 = pd.read_table(count_Test1 , index_col = 0 , skiprows = 1)\n","\n","  # TPMの計算につかわない列を削除\n","  drop_col = ['Chr' , 'Start' , 'End' , 'Strand']\n","  df_Cont1 = df_Cont1.drop(drop_col, axis=1)\n","\n","  drop_col = ['Chr' , 'Start' , 'End' , 'Strand' , 'Length']\n","  df_Test1 = df_Test1.drop(drop_col, axis=1)\n","\n","  # 表をくっつける、名前の変更\n","  df_Merge = pd.merge(df_Cont1 , df_Test1, on='Geneid')\n","  df_Merge.columns = ['Length',\n","                      Cont_name ,\n","                      Test_name]\n","\n","# N=3の処理\n","elif n_number == \"3\":\n","    # データの読み込み\n","    df_Cont1 = pd.read_table(count_Cont1 , index_col = 0 , skiprows = 1)\n","    df_Cont2 = pd.read_table(count_Cont2 , index_col = 0 , skiprows = 1)\n","    df_Cont3 = pd.read_table(count_Cont3 , index_col = 0 , skiprows = 1)\n","    df_Test1 = pd.read_table(count_Test1 , index_col = 0 , skiprows = 1)\n","    df_Test2 = pd.read_table(count_Test2 , index_col = 0 , skiprows = 1)\n","    df_Test3 = pd.read_table(count_Test3 , index_col = 0 , skiprows = 1)\n","\n","    # TPMの計算につかわない列を削除\n","    drop_col = ['Chr' , 'Start' , 'End' , 'Strand']\n","    df_Cont1 = df_Cont1.drop(drop_col, axis=1)\n","\n","    drop_col = ['Chr' , 'Start' , 'End' , 'Strand' , 'Length']\n","    df_Cont2 = df_Cont2.drop(drop_col, axis=1)\n","    df_Cont3 = df_Cont3.drop(drop_col, axis=1)\n","    df_Test1 = df_Test1.drop(drop_col, axis=1)\n","    df_Test2 = df_Test2.drop(drop_col, axis=1)\n","    df_Test3 = df_Test3.drop(drop_col, axis=1)\n","\n","    # 表をくっつける、名前の変更\n","    df_Merge = pd.merge(df_Cont1, df_Cont2, on='Geneid')\n","    df_Merge = pd.merge(df_Merge, df_Cont3, on='Geneid')\n","    df_Merge = pd.merge(df_Merge, df_Test1, on='Geneid')\n","    df_Merge = pd.merge(df_Merge, df_Test2, on='Geneid')\n","    df_Merge = pd.merge(df_Merge, df_Test3, on='Geneid')\n","\n","    df_Merge.columns = ['Length',\n","                        Cont_name + \"_1\" ,\n","                        Cont_name + \"_2\" ,\n","                        Cont_name + \"_3\" ,\n","                        Test_name + \"_1\" ,\n","                        Test_name + \"_2\" ,\n","                        Test_name + \"_3\"]\n","\n","#以下共通、出力する\n","# df_Merge.to_csv(exp_name + \"countdata.csv\")\n","# print(df_Merge)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"8TWrwRUBHh5n","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1638250436695,"user_tz":-540,"elapsed":444,"user":{"displayName":"久内佑樹","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gg8aqObfU0mdL-poY9eo71yADwG-v3GfTCsl8fn=s64","userId":"01891010281923395733"}},"outputId":"6aac3977-38b8-4aac-97d6-93fb5c0821b1"},"source":["#@title 調整したカウントデータとid_nameの結合\n","#@markdown Gene_idを元に、カウントデータにGene_nameを結合する\n","\n","# merge(左側のDataFrame, 右側のDataFrame , on=どの行を見るか)で結合\n","df_with_product = pd.merge(df_id_name , df_Merge , on='Geneid')\n","\n","# インデックス行を指定\n","df_with_product = df_with_product.set_index([\"Geneid\" , \"product\"])\n","\n","print(df_with_product)"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["                            Length  N_DC_1  N_DC_2  ...  H_DC_1  H_DC_2  H_DC_3\n","Geneid          product                             ...                        \n","ENSG00000000003 TSPAN6        4535     719     594  ...     635     612     804\n","ENSG00000000005 TNMD          1610       0       0  ...       0       0       0\n","ENSG00000000419 DPM1          1207     686     612  ...     725     796     817\n","ENSG00000000457 SCYL3         6883     209     179  ...     182     248     200\n","ENSG00000000460 C1orf112      5967      86     163  ...      67      98      77\n","...                            ...     ...     ...  ...     ...     ...     ...\n","ENSG00000285990 AL589743.7     647       0       0  ...       0       0       0\n","ENSG00000285991 AL355312.5    5065       1       0  ...       0       0       0\n","ENSG00000285992 AC120036.5     956       0       0  ...       0       0       0\n","ENSG00000285993 AC018931.1    1246       0       0  ...       0       0       0\n","ENSG00000285994 AL731559.1    3732       0       0  ...       0       0       0\n","\n","[58676 rows x 7 columns]\n"]}]},{"cell_type":"markdown","metadata":{"id":"bQePf0AdLrVh"},"source":["===================================================================\n","============================================"]},{"cell_type":"code","metadata":{"id":"Yt-A19XrLsQK"},"source":["#@title iDEP用のidとカウントを対応させたcsvを出力させる\n","df_iDEP = df_with_product.reset_index()\n","df_iDEP = df_iDEP.set_index(\"Geneid\")\n","\n","df_iDEP = df_iDEP.drop([\"product\" , \"Length\"] , axis = 1)\n","df_iDEP.to_csv(exp_name + \"for_iDEP.csv\", sep=\",\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"1_CEP44Cbrtm"},"source":["===================================================================\n","============================================"]},{"cell_type":"code","metadata":{"id":"TUyzPImxH_os"},"source":["#@title TPMの計算式を定義\n","\n","# 計算式を定義しているだけ\n","\n","def normalize_per_million_reads(df):\n","    # RPM/FPM = reads per million  fragments per million\n","    sum = df.sum()\n","    return 10**6 * df / sum    # 列ごとにカウント/総リード数計算\n","\n","def normalize_per_kilobase(df, gene_length):\n","    # FPKM = fragments per kilobase of exon per million reads mapped\n","    ret = ((df.copy().T) * 10**3 / gene_length).T\n","    return ret\n","\n","def normalize_tpm(df, gene_length):\n","    # TPM = transcripts per kilobase million \n","    # normalize_per_kilobaseの結果にnormalize_per_millionを施す\n","    df_tmp = df.copy()\n","    df_tmp = normalize_per_kilobase(df_tmp, gene_length)\n","    df_tmp = normalize_per_million_reads(df_tmp)\n","    return df_tmp"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"2IYLYh0mIBev"},"source":["#@title カウントデータをTPM正規化\n","\n","if n_number == \"1\":\n","  df_count = df_with_product.iloc[: , 1:3]\n","\n","if n_number == \"3\":\n","  df_count = df_with_product.iloc[: , 1:7]\n","\n","\n","gene_length = df_with_product['Length'] \n","\n","# TPMによる正規化\n","df_tpm = normalize_tpm(df_count, gene_length)\n","df_tpm.to_csv(exp_name + \"tpm.csv\", sep=\",\")\n","\n","# print(df_tpm)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"5k9E7cGGbtEO"},"source":["===================================================================\n","============================================"]},{"cell_type":"markdown","metadata":{"id":"PNm4ptIebwQY"},"source":["出力ファイル<br>\n","・実験名_id_name.csv\n","　　　　→Gene_idとProduct(遺伝子名)との対応表<br>\n","・実験名_for_iDEP.csv\n","　　　　→Gene_idとカウント数との対応表<br>\n","・実験名_tpm.csv\n","　　　　　　→遺伝子名とTPM値を対応付けたcsvファイル"]}]}